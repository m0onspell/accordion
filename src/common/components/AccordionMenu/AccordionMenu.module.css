.container {
  padding: 0;
  list-style: none;
  margin: 1.25rem auto;
  box-shadow: 0 0 25px #00000070;
}
