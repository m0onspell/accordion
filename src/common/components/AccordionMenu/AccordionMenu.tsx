import React, { FC, memo } from 'react';
import styles from './AccordionMenu.module.css';
import { useLocation } from 'react-router-dom';
import { GroupMenuItem } from './GroupMenuItem';
import { MenuItem } from './MenuItem';

export type MenuItemProps = {
  label: string;
  href: string;
  items?: MenuItemProps[];
};

export interface AccordionMenuProps {
  className?: string;
  items: MenuItemProps[];
}

const mapItems = (items: MenuItemProps[], currentPathname: string, basePath: string = '') =>
  items.map(({ label, href, items }) => {
    const fullPath = `${basePath}${href}`;
    const isActive = currentPathname.startsWith(fullPath);
    return items ? (
      <GroupMenuItem isActive={isActive} label={label} key={fullPath}>
        {mapItems(items, currentPathname, fullPath)}
      </GroupMenuItem>
    ) : (
      <MenuItem label={label} href={fullPath} isActive={isActive} key={fullPath} />
    );
  });

const AccordionMenu: FC<AccordionMenuProps> = memo(({ className, items }) => {
  const { pathname } = useLocation();
  return (
    <nav className={className}>
      <ul className={styles.container}>{mapItems(items, pathname)}</ul>
    </nav>
  );
});

export { AccordionMenu };
