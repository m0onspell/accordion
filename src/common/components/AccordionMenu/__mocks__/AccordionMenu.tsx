import React, { FC } from 'react';

const AccordionMenu: FC = () => {
  return <>AccordionMenu</>;
};

export { AccordionMenu };
