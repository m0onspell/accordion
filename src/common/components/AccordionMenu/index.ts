import { AccordionMenu, MenuItemProps as MenuItem } from './AccordionMenu';

export { AccordionMenu };

export type MenuItemProps = MenuItem;
