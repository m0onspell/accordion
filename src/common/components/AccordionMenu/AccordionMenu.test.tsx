import React from 'react';
import renderer from 'react-test-renderer';
import { AccordionMenu } from './AccordionMenu';
import { MemoryRouter } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { AccordionMenuProps } from './AccordionMenu';
import * as H from 'history';

const makeProps = (props: Partial<AccordionMenuProps> = {}): AccordionMenuProps => ({
  items: [
    {
      label: 'Activities',
      href: '/activities',
      items: [
        { label: 'Recreational', href: '/recreational' },
        {
          label: 'Cooking',
          href: '/cooking',
        },
      ],
    },
    {
      label: 'Account',
      href: '/account',
      items: [
        { label: 'Profile', href: '/profile' },
        { label: 'Settings', href: '/settings' },
      ],
    },
  ],
  className: 'testClass',
  ...props,
});

const setup = ({
  props,
  initialEntries,
}: { props?: Partial<AccordionMenuProps>; initialEntries?: H.LocationDescriptor[] } = {}) => (
  <MemoryRouter initialEntries={initialEntries}>
    <AccordionMenu {...makeProps(props)} />
  </MemoryRouter>
);

test('renders correctly', () => {
  const tree = renderer.create(setup({ initialEntries: ['/activities/cooking'] })).toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders with current route menu link active', () => {
  render(setup({ initialEntries: ['/account/profile'] }));
  expect(screen.getByText(/Account/).parentElement).toHaveClass('activeSection');
  expect(screen.getByText(/Profile/)).toHaveClass('activeLink');
});

test('click on a section title toggles item visibility while keeping previous section state', () => {
  render(setup({ initialEntries: ['/account/profile'] }));
  const activities = screen.getByText(/Activities/);
  const account = screen.getByText(/Account/);

  expect(account.parentElement).toHaveClass('activeSection');

  userEvent.click(activities);

  expect(account.parentElement).toHaveClass('activeSection');
  expect(activities.parentElement).toHaveClass('activeSection');

  userEvent.click(activities);

  expect(account.parentElement).toHaveClass('activeSection');
  expect(activities.parentElement).not.toHaveClass('activeSection');
});
