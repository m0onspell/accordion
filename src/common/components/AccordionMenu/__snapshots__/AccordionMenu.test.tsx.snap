// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<nav
  className="testClass"
>
  <ul
    className="container"
  >
    <li
      className="item activeSection"
    >
      <div
        aria-expanded={true}
        aria-haspopup="true"
        className="sectionTitle"
        onClick={[Function]}
        role="button"
      >
        Activities
         
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-angle-up fa-w-10 icon"
          data-icon="angle-up"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M177 159.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 255.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 329.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
      <ul
        aria-hidden={false}
        aria-label="Activities submenu"
        className="subItems"
      >
        <li
          className="item"
        >
          <a
            className="menuLink"
            href="/activities/recreational"
            onClick={[Function]}
          >
            Recreational
          </a>
        </li>
        <li
          className="item"
        >
          <a
            className="menuLink activeLink"
            href="/activities/cooking"
            onClick={[Function]}
          >
            Cooking
          </a>
        </li>
      </ul>
    </li>
    <li
      className="item"
    >
      <div
        aria-expanded={false}
        aria-haspopup="true"
        className="sectionTitle"
        onClick={[Function]}
        role="button"
      >
        Account
         
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-angle-down fa-w-10 icon"
          data-icon="angle-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
      <ul
        aria-hidden={true}
        aria-label="Account submenu"
        className="subItems"
      >
        <li
          className="item"
        >
          <a
            className="menuLink"
            href="/account/profile"
            onClick={[Function]}
          >
            Profile
          </a>
        </li>
        <li
          className="item"
        >
          <a
            className="menuLink"
            href="/account/settings"
            onClick={[Function]}
          >
            Settings
          </a>
        </li>
      </ul>
    </li>
  </ul>
</nav>
`;
